discord/client.py

```diff
@@ -1155,6 +1158,53 @@ def transform(d):
                        user=User(data=user, state=state),
                        connected_accounts=data['connected_accounts'])
 
+    async def fetch_channel(self, channel_id):
+        """|coro|
+
+        Retrieves a :class:`.abc.GuildChannel` or :class:`.abc.PrivateChannel` with the specified ID.
+
+        .. note::
+
+            This method is an API call. For general usage, consider :meth:`get_channel` instead.
+
+        .. versionadded:: 1.2.0
+
+        Raises
+        -------
+        TypeError
+            An unknown channel type was received from Discord.
+        :exc:`.HTTPException`
+            Retrieving the channel failed.
+        :exc:`.NotFound`
+            Invalid Channel ID.
+        :exc:`.Forbidden`
+            You do not have permission to fetch this channel.
+
+        Returns
+        --------
+        Union[:class:`.abc.GuildChannel`, :class:`.abc.PrivateChannel`]
+            The channel from the ID.
+        """
+        data = await self.http.get_channel(channel_id)
+
+        factory, ch_type = _channel_factory(data['type'])
+        if factory is None:
+            raise TypeError('Unknown channel type %s.' % data['type'])
+
+        if ch_type in (ChannelType.group, ChannelType.private):
+            channel = factory(me=self.user, data=data, state=self._connection)
+        else:
+            guild_id = utils._get_as_snowflake(data, 'guild_id')
+            guild = self.get_guild(guild_id)
```

Comment:
```suggestion
            guild = self.get_guild(guild_id) or Object(id=guild_id)
```

Might as well do this in one go.

Note that the documentation needs to mention this quirk in the implementation.